varcode/variant.py

```diff
@@ -46,3 +114,240 @@ def short_description(self):
+
+    def coding_transcripts(self):
+        """
+        Protein coding transcripts
+        """
+        return [
+            transcript for transcript in self.transcripts()
+            if is_coding_biotype(transcript.biotype)
+        ]
+
+    def genes(self):
+        return self.ensembl.genes_at_locus(
+            self.contig, self.pos, self.end)
+
+    def coding_genes(self):
+        """
+        Protein coding transcripts
+        """
+        return [
+            gene for gene in self.genes()
+            if is_coding_biotype(gene.biotype)
+        ]
+
+    def annotate(self, raise_on_error=True):
+        """
+        Determine the effects of a variant on any transcripts it overlaps.
+        Returns a Annotation object.
+        """
+
+        overlapping_transcripts = self.transcripts()
+
+        if len(overlapping_transcripts) == 0:
+            # intergenic variant
+            return Annotation(
+                variant=self,
+                genes=[],
+                gene_transcript_effects={},
+                errors={})
+
+        # group transcripts by their gene ID
+        overlapping_transcript_groups = group_by(
+            overlapping_transcripts, field_name='gene_id')
+
+        # dictionary from gene ID to list of transcript effects
+        gene_transcript_effects_groups = {}
+
+        # mapping from Transcript objects to errors encountered
+        # while trying to annotated them
+        errors = {}
+
+        for (gene_id, transcripts) in overlapping_transcript_groups.items():
+            effects = []
+            for transcript in transcripts:
+                try:
+                    effects.append(self.transcript_effect(transcript))
+                except (AssertionError, ValueError) as error:
+                    if raise_on_error:
+                        raise
+                    else:
+                        logging.warn(
+                            "Encountered error annotating %s for %s: %s",
+                            self,
+                            transcript,
+                            error)
+                    errors[transcript] = error
+            gene_transcript_effects_groups[gene_id] = effects
+
+        # construct Gene objects for all the genes containing
+        # all the transcripts this variant overlaps
+        overlapping_genes = [
+            self.ensembl.gene_by_id(gene_id)
+            for gene_id
+            in overlapping_transcript_groups.keys()
+        ]
+
+        return Annotation(
+            variant=self,
+            genes=overlapping_genes,
+            gene_transcript_effects=gene_transcript_effects_groups,
+            errors=errors)
+
+    def transcript_effect(self, transcript):
+        """
+        Generate a transcript effect (such as FrameShift) by applying a
+        genomic variant to a particular transcript.
+
+        Parameters
+        ----------
+
+        transcript :  Transcript
+            Transcript we're going to mutate
```

Comment:
my first reading of the word "mutate" was in the software sense, not the biological sense. Might change this to the more wordy but clearer "Transcript we're going to apply the mutation to". May also want to be clear about this by changing the main method description above to something like "Return the transcript effect (such as FrameShift) from applying a genomic variant to a particular transcript."